# Contents

###  [1. Variance Thershold](#1.-Variance-Threshold)


In [1]:
# creating a data frame 

import pandas as pd

df = pd.DataFrame({'A':[1,2,4,1,2,4],
                   'B':[4,5,6,7,8,9],
                   'C':[0,0,0,0,0,0],
                   'D':[1,1,1,1,1,1]})

# 1. Variance Threshold
Feature selector that removes all low-variance features.

This feature selection algorithm looks only at the features (X), not the desired outputs (y), and can thus be used for unsupervised learning.

## Dropping constant features
- Removing those features that has only constant values or zero variance as it of no use for solving the problem statement.

In [4]:
from sklearn.feature_selection import VarianceThreshold
varthes = VarianceThreshold(threshold = 0)
varthes.fit(df)

VarianceThreshold(threshold=0)

In [5]:
varthes.get_support()

array([ True,  True, False, False])

In [8]:
constant_columns = [column for column in df.columns
                    if column not in df.columns[varthes.get_support()]]
print(constant_columns)

['C', 'D']


- A and B features are non zero variance features and therefore they are usefully in solving a problem statement.
- C and B features are zero variance features and therefore they can be removed.

In [10]:
# removing zero variance features

df.drop(constant_columns, axis=1, inplace=True)

In [11]:
df

A  B
0  1  4
1  2  5
2  4  6
3  1  7
4  2  8
5  4  9

## Working with kaggle dataset

We are gonna play with 'Santander Customer Satisfaction' dataset from kaggle.  
There are two file 'train.csv' and 'test.csv' .  
The dataset contains totally 371 features so lets get to work and start selecting only the useful features.   
The dependent feature is 'TARGET'. Rest all are independent features.    

In [12]:
# importing the dataset

df_train = pd.read_csv(r'F:\Data Science\Datasets\Santander Customer Satisfaction\train.csv')
df_train.head()

ID  var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
0   1     2     23                 0.0                      0.0   
1   3     2     34                 0.0                      0.0   
2   4     2     23                 0.0                      0.0   
3   8     2     37                 0.0                    195.0   
4  10     2     39                 0.0                      0.0   

   imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                    195.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   imp_op_var40_efect_ult1  imp_op_var40_efect_ult3  ...  \
0                      0.0                      0.0  ...   
1                      0.0                      0.0  ...   
2                      0.0                      0.0  ...   
3                      0.0                      0.0  ...   
4                      0.0                      0.0  ...   

   saldo_medio_var33_hace2  saldo_medio_var33_hace3  saldo_medio_var33_ult1  \
0                      0.0                      0.0                     0.0   
1                      0.0                      0.0                     0.0   
2                      0.0                      0.0                     0.0   
3                      0.0                      0.0                     0.0   
4                      0.0                      0.0                     0.0   

   saldo_medio_var33_ult3  saldo_medio_var44_hace2  saldo_medio_var44_hace3  \
0                     0.0                      0.0                      0.0   
1                     0.0                      0.0                      0.0   
2                     0.0                      0.0                      0.0   
3                     0.0                      0.0                      0.0   
4                     0.0                      0.0                      0.0   

   saldo_medio_var44_ult1  saldo_medio_var44_ult3       var38  TARGET  
0                     0.0                     0.0   39205.170       0  
1                     0.0                     0.0   49278.030       0  
2                     0.0                     0.0   67333.770       0  
3                     0.0                     0.0   64007.970       0  
4                     0.0                     0.0  117310.979       0  

[5 rows x 371 columns]

In [13]:
df_train.shape

(76020, 371)

The dataset contains around 76,000 rows and 371 columns.

In [31]:
# separating dependent and independent features

x = df_train.drop('TARGET', axis = 1)
y = df_train['TARGET']

In [39]:
x.head()

ID  var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
0   1     2     23                 0.0                      0.0   
1   3     2     34                 0.0                      0.0   
2   4     2     23                 0.0                      0.0   
3   8     2     37                 0.0                    195.0   
4  10     2     39                 0.0                      0.0   

   imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                    195.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   imp_op_var40_efect_ult1  imp_op_var40_efect_ult3  ...  \
0                      0.0                      0.0  ...   
1                      0.0                      0.0  ...   
2                      0.0                      0.0  ...   
3                      0.0                      0.0  ...   
4                      0.0                      0.0  ...   

   saldo_medio_var29_ult3  saldo_medio_var33_hace2  saldo_medio_var33_hace3  \
0                     0.0                      0.0                      0.0   
1                     0.0                      0.0                      0.0   
2                     0.0                      0.0                      0.0   
3                     0.0                      0.0                      0.0   
4                     0.0                      0.0                      0.0   

   saldo_medio_var33_ult1  saldo_medio_var33_ult3  saldo_medio_var44_hace2  \
0                     0.0                     0.0                      0.0   
1                     0.0                     0.0                      0.0   
2                     0.0                     0.0                      0.0   
3                     0.0                     0.0                      0.0   
4                     0.0                     0.0                      0.0   

   saldo_medio_var44_hace3  saldo_medio_var44_ult1  saldo_medio_var44_ult3  \
0                      0.0                     0.0                     0.0   
1                      0.0                     0.0                     0.0   
2                      0.0                     0.0                     0.0   
3                      0.0                     0.0                     0.0   
4                      0.0                     0.0                     0.0   

        var38  
0   39205.170  
1   49278.030  
2   67333.770  
3   64007.970  
4  117310.979  

[5 rows x 370 columns]

In [41]:
y.head()

0    0
1    0
2    0
3    0
4    0
Name: TARGET, dtype: int64

In [42]:
# train, test, split

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.3, random_state = 40)

In [43]:
# applying varians threshold

varthes.fit(x_train)

VarianceThreshold(threshold=0)

In [47]:
varthes.get_support()

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True, False, False,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True, False, False, False, False,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
       False, False,  True,  True,  True,  True,  True,  True,  True,
       False,  True,  True,  True, False, False,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,

In [46]:
# non constant features

sum(varthes.get_support())

327

There are 327 features that are non-constant from the total of 370 features.  
So 43 features are zero variance features.  

In [49]:
# names of the zero variance / constant features

constant_columns1 = [columns for columns in x_train.columns
                     if columns not in x_train.columns[varthes.get_support()]]
constant_columns1

['ind_var2_0',
 'ind_var2',
 'ind_var27_0',
 'ind_var28_0',
 'ind_var28',
 'ind_var27',
 'ind_var34_0',
 'ind_var34',
 'ind_var41',
 'ind_var46_0',
 'ind_var46',
 'num_var27_0',
 'num_var28_0',
 'num_var28',
 'num_var27',
 'num_var34_0',
 'num_var34',
 'num_var41',
 'num_var46_0',
 'num_var46',
 'saldo_var28',
 'saldo_var27',
 'saldo_var34',
 'saldo_var41',
 'saldo_var46',
 'delta_imp_amort_var34_1y3',
 'imp_amort_var18_hace3',
 'imp_amort_var34_hace3',
 'imp_amort_var34_ult1',
 'imp_reemb_var13_hace3',
 'imp_reemb_var17_hace3',
 'imp_reemb_var33_hace3',
 'imp_trasp_var17_out_hace3',
 'imp_trasp_var33_out_hace3',
 'num_var2_0_ult1',
 'num_var2_ult1',
 'num_reemb_var13_hace3',
 'num_reemb_var17_hace3',
 'num_reemb_var33_hace3',
 'num_trasp_var17_out_hace3',
 'num_trasp_var33_out_hace3',
 'saldo_var2_ult1',
 'saldo_medio_var13_medio_hace3']

In [50]:
print(len(constant_columns1))

43


We can see that there are totally 43 constant features and we have to remove those.

In [51]:
# removing constant features

x_train.drop(constant_columns1, axis = 1)

ID  var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
11453   22966     2     33                 0.0                      0.0   
4090     8196     2     32                 0.0                      0.0   
57552  114859     2     29                 0.0                      0.0   
49183   98240     2     60                 0.0                      0.0   
27692   55426     2     23                 0.0                      0.0   
...       ...   ...    ...                 ...                      ...   
11532   23110     2     39                 0.0                   1800.0   
54578  108880     2     46                 0.0                      0.0   
56760  113246     2     48                 0.0                      0.0   
30727   61390     2     30                 0.0                      0.0   
11590   23232     2     25                 0.0                      0.0   

       imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  \
11453                     0.00                      0.0   
4090                      0.00                      0.0   
57552                     0.00                      0.0   
49183                     0.00                      0.0   
27692                     0.00                      0.0   
...                        ...                      ...   
11532                  2009.13                      0.0   
54578                     0.00                      0.0   
56760                     0.00                      0.0   
30727                     0.00                      0.0   
11590                     0.00                      0.0   

       imp_op_var40_comer_ult3  imp_op_var40_efect_ult1  \
11453                      0.0                      0.0   
4090                       0.0                      0.0   
57552                      0.0                      0.0   
49183                      0.0                      0.0   
27692                      0.0                      0.0   
...                        ...                      ...   
11532                      0.0                      0.0   
54578                      0.0                      0.0   
56760                      0.0                      0.0   
30727                      0.0                      0.0   
11590                      0.0                      0.0   

       imp_op_var40_efect_ult3  ...  saldo_medio_var29_ult3  \
11453                      0.0  ...                     0.0   
4090                       0.0  ...                     0.0   
57552                      0.0  ...                     0.0   
49183                      0.0  ...                     0.0   
27692                      0.0  ...                     0.0   
...                        ...  ...                     ...   
11532                      0.0  ...                     0.0   
54578                      0.0  ...                     0.0   
56760                      0.0  ...                     0.0   
30727                      0.0  ...                     0.0   
11590                      0.0  ...                     0.0   

       saldo_medio_var33_hace2  saldo_medio_var33_hace3  \
11453                      0.0                      0.0   
4090                       0.0                      0.0   
57552                      0.0                      0.0   
49183                      0.0                      0.0   
27692                      0.0                      0.0   
...                        ...                      ...   
11532                      0.0                      0.0   
54578                      0.0                      0.0   
56760                      0.0                      0.0   
30727                      0.0                      0.0   
11590                      0.0                      0.0   

       saldo_medio_var33_ult1  saldo_medio_var33_ult3  \
11453                     0.0                     0.0   
4090                      0.0                     0.0   
57552                     0.0                     0.0  

Now we have totally 327 useful features for solving our problem statement.  
We can also change the variance threshold from 0 to 5 or 10, by doing so will reduce the total number of constant features.

# -----------------------------------------------------------------------------------------------------------------